xdsl/traits.py

```diff
@@ -444,30 +445,98 @@ def get_canonicalization_patterns(cls) -> tuple[RewritePattern, ...]:
         raise NotImplementedError()
 
 
+class MemoryEffectKind(Enum):
+    """
+    The kind of side effect an operation can have.
+
+    MLIR has a more detailed version of this, able to tie effects to specfic resources or
+    values. Here, everything has its effect on the universe.
+    """
+
+    READ = auto()
+    """
+    Indicates that the operation reads from some resource. A 'read' effect implies only
+    dereferencing of the resource, and not any visible mutation.
+    """
+
+    WRITE = auto()
+    """
+    Indicates that the operation writes to some resource. A 'write' effect implies only
+    mutating a resource, and not any visible dereference or read.
+    """
+
+    ALLOC = auto()
+    """
+    Indicates that the operation allocates from some resource. An 'allocate' effect
+    implies only allocation of the resource, and not any visible mutation or dereference.
+    """
+
+    FREE = auto()
+    """
+    Indicates that the operation frees some resource that has been allocated. A 'free'
+    effect implies only de-allocation of the resource, and not any visible allocation,
+    mutation or dereference.
+    """
+
+
 class MemoryEffect(OpTrait):
     """
     A trait that enables operations to expose their side-effects or absence thereof.
-
-    NB: The MLIR implementation further allows to describe what *kind* of side-effects
-    an operation has, e.g., read-only, or allocation.
-    This one is a stripped down version for now, just saying if there are any
-    side-effects or not.
     """
 
     @classmethod
     @abc.abstractmethod
-    def has_effects(cls, op: Operation) -> bool:
+    def get_effects(cls, op: Operation) -> set[MemoryEffectKind] | None:
+        """
+        Returns the concrete side effects of the operation.
+
+        Return None if the operation cannot conclude - interpreted as if the operation
+        had no MemoryEffect interface in the first place.
+        """
         raise NotImplementedError()
 
+    @final
+    @classmethod
+    def has_effects(cls, op: Operation) -> bool:
+        """
+        Returns if the operation has any side effects.
+        """
+        effects = cls.get_effects(op)
+        return (effects is not None) and len(effects) > 0
+
 
-def is_side_effect_free(op: Operation):
+def only_has_effect(op: Operation, effect: MemoryEffectKind) -> bool:
+    """
+    Returns if the operation has the given side effects and no others.
+    """
+    return get_effects(op) == {effect}
+
+
+def is_side_effect_free(op: Operation) -> bool:
     """
     Boilerplate helper to check if a generic operation is side effect free for sure.
     """
-    # If it doesn't say, safely assume it has side effects.
-    if not (trait := op.get_trait(MemoryEffect)):
-        return False
-    return not trait.has_effects(op)
+    return get_effects(op) == set()
```

Comment:
nit:
```suggestion
    effects = get_effects(op)
    return effects is not None and bool(effects)
```
Avoids the empty set construction

Reply:
Is Python really not optimizing this?
Your proposal is off but I can get it to a similar form if you really want to.

I'm wondering if this is not the kind of things we should just keep readable and just hope the most basic optimization gets rid of though?